### PR TITLE
Add `test_run_starting` event to broadcast all test cases before a run

### PR DESCRIPTION
**Pull request by mattwynne**
Any listener that wants to know how many test cases there will be (e.g. to display a progress bar) will need this information.

### CODE REVIEW

**Comment by mattwynne** [comment]
Part of #962

**Comment by brasmusson** [comment]
Is it really a good idea that because one listener needs to know the number test cases, also when that listener is not used, the first test case cannot be executed before all feature files have been parsed and compiled? Isn't this a major change from the original design of "the new core". Shouldn't this be subject of a command line option?

File: lib/cucumber/events.rb [hunk 1]
```diff
@@ -32,6 +32,7 @@ def self.registry
         StepActivated,
         TestRunFinished,
         GherkinSourceRead
+        TestRunStarting,
```
Comment (olleolleolle) [inline lib/cucumber/events.rb]: The comma seems to be misplaced?

**Comment by brasmusson** [comment]
Also I think that an event named "TestRunStarting" should be issued first before any feature files has been read, and that the time printed in the summary should be the time difference between the time stamp of the TestRunStaring event and the time stamp of the TestRunFinished event.

**Comment by mattwynne** [comment]
*now* you tell me 😜

@brassmusson good point. maybe we could be clever enough to only add this filter if there is anyone listening to the event?

File: lib/cucumber/events.rb (hunk 1 above)
Comment (olleolleolle) [inline lib/cucumber/events.rb]: https://github.com/cucumber/cucumber-ruby/pull/1124

**Pull request merged**